src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlerTests.IgnoreCycle.cs

```diff
@@ -0,0 +1,324 @@
+        public async Task AlreadySeenInstance_ShouldNotBeIgnoredOnSecondBranch(Type objectType, string objectPayload)
+        {
+            object obj = Activator.CreateInstance(objectType);
+            var root = new TreeNode();
+            root.Left = obj;
+            root.Right = obj;
+
+            await Test_Serialize_And_SerializeAsync(root, $@"{{""Left"":{objectPayload},""Right"":{objectPayload}}}", s_optionsIgnoreCycles);
+        }
+
+        [Fact]
+        public async Task IgnoreCycles_WhenWritingNull()
+        {
+            var opts = new JsonSerializerOptions
+            {
+                ReferenceHandler = ReferenceHandler.IgnoreCycle,
+                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
+            };
+
+            // Reference cycles are treated as null, hence the JsonIgnoreCondition can be used to actually ignore the property.
+            var rootObj = new Node();
+            rootObj.Next = rootObj;
+
+            await Test_Serialize_And_SerializeAsync(rootObj, "{}", opts);
+
+            // JsonIgnoreCondition does not ignore nulls in collections, hence a reference loop should not omit writing it.
+            // This also helps us to avoid changing the length of a collection when the loop is detected in one of the elements.
+            var rootList = new List<object>();
+            rootList.Add(rootList);
+
+            await Test_Serialize_And_SerializeAsync(rootList, "[null]", opts);
+
+            var rootDictionary = new Dictionary<string, object>();
+            rootDictionary.Add("self", rootDictionary);
+
+            await Test_Serialize_And_SerializeAsync(rootDictionary, @"{""self"":null}", opts);
+        }
+
+        private async Task Test_Serialize_And_SerializeAsync(object obj, string expected, JsonSerializerOptions options)
+        {
+            string json = JsonSerializer.Serialize(obj, options);
+            Assert.Equal(expected, json);
+
+            using var ms = new MemoryStream();
+            await JsonSerializer.SerializeAsync(ms, obj, options).ConfigureAwait(false);
+            json = Encoding.UTF8.GetString(ms.ToArray());
+            Assert.Equal(expected, json);
+        }
+
+        private class Employee
+        {
+            public string Name { get; set; }
+            public Employee Manager { get; set; }
+            public List<Employee> Subordinates { get; set; }
+            public Dictionary<string, Employee> Colleagues { get; set; }
+        }
+
+        private class Node
+        {
+            public object Next { get; set; }
+        }
+
+        private class TreeNode
+        {
+            public object Left { get; set; }
+            public object Right { get; set; }
+        }
+
+        private struct ValueNode : IValueNode
+        {
+            public object Next { get; set; }
+        }
+
+        interface IValueNode
+        {
+            public object Next { get; set; }
+        }
```

Comment:
All of these types have `object`-typed properties, and go through the polymorphic code paths. Can we test variants of these types with strongly typed class-type, value-type, and nullable-value-type properties?